### PR TITLE
Code generation

### PR DESCRIPTION
**Pull request by maelle**
cf #41

### CODE REVIEW

File: content/post/2020-02-04-code-generation/index.Rmd [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+See [aaa.R](https://github.com/coolbutuseless/minicss/blob/fe378f6e040405e51fb07cb74fd2f0bac8a85b26/R/aaa.R) and [prop_transform.R](https://github.com/coolbutuseless/minicss/blob/fe378f6e040405e51fb07cb74fd2f0bac8a85b26/R/prop_transform.R). 
+As in most examples the code is generated as a string, but in that case it's not written to disk, it becomes code via the use of [`eval()` and `parse()`](http://adv-r.had.co.nz/Expressions.html#parsing-and-deparsing).
+
+#### Generate C++ bindings with `Rcpp::compileAttributes()`
+
+`Rcpp::compileAttributes()` generates code (the bindings required to call C++ functions from R) after scanning a package source files. Find more information [in the `Rcpp` vignette about attributes](https://cran.r-project.org/web/packages/Rcpp/vignettes/Rcpp-attributes.pdf). You could call the function "whenever functions areadded, removed, or have their signatures changed." but the aforementioned vignette also states " if you are using either RStudio or `devtools`to build your package then the `compileAttributes` function is called automatically whenever your package is built".
```
Comment (MilesMcBain) [inline content/post/2020-02-04-code-generation/index.Rmd]: areadded -> are read?

**Comment by MilesMcBain** [comment]
I LOVE THIS. This reminds of some of those glorious summary papers out there in academia land that collect together a bunch of the best papers on a topic. Stumbling on one of those always felts like striking gold and I get the same feeling from reading this. ⭐️ ⭐️ ⭐️ ⭐️ ⭐️

File: content/post/2020-02-04-code-generation/index.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+See [aaa.R](https://github.com/coolbutuseless/minicss/blob/fe378f6e040405e51fb07cb74fd2f0bac8a85b26/R/aaa.R) and [prop_transform.R](https://github.com/coolbutuseless/minicss/blob/fe378f6e040405e51fb07cb74fd2f0bac8a85b26/R/prop_transform.R). 
+As in most examples the code is generated as a string, but in that case it's not written to disk, it becomes code via the use of [`eval()` and `parse()`](http://adv-r.had.co.nz/Expressions.html#parsing-and-deparsing).
+
+#### Generate C++ bindings with `Rcpp::compileAttributes()`
+
+`Rcpp::compileAttributes()` generates code (the bindings required to call C++ functions from R) after scanning a package source files. Find more information [in the `Rcpp` vignette about attributes](https://cran.r-project.org/web/packages/Rcpp/vignettes/Rcpp-attributes.pdf). You could call the function "whenever functions areadded, removed, or have their signatures changed." but the aforementioned vignette also states " if you are using either RStudio or `devtools`to build your package then the `compileAttributes` function is called automatically whenever your package is built".
```
Comment (gaborcsardi) [inline content/post/2020-02-04-code-generation/index.md]: areadded -> are added

Comment (gaborcsardi) [inline content/post/2020-02-04-code-generation/index.md]: scrap space after quote in `states " if you`

File: content/post/2020-02-04-code-generation/index.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+`Rcpp::compileAttributes()` is with `scaffolder` one of the only "general" solutions mentioned in this post, as opposed as other approaches developed for and within a given package.
```
Comment (gaborcsardi) [inline content/post/2020-02-04-code-generation/index.md]: I am not really sure what this means, especially the first part.

File: content/post/2020-02-04-code-generation/index.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+### Generating code on-the-fly
+
+One step further, one might generate code on-the-fly, i.e. as users run the package.
+
+* The [`chromote`](https://github.com/rstudio/chromote) package ["generates auto-completable R6 methods at runtime"](https://twitter.com/alandipert/status/1199456564442062848).
```
Comment (gaborcsardi) [inline content/post/2020-02-04-code-generation/index.md]: What does "on-the-fly" mean? When you load the package? When you run a function? When you create an R6 class? When you create an R6 object?

File: content/post/2020-02-04-code-generation/index.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+_Thanks to Rich for many useful comments on this post._
+
+## Conclusion
+
+In this post we explored different aspects of source code scaffolding and generation in R packages. 
+We've mentioned examples of code scaffolding (`gitea`, `scaffolder`), of code generation by a script (`wisegroup`, `eml.build`, `redux`, `xaringanthemer`) or by a meta package (`vlbuildr` and `vlmetabuildr`) before package shipping, of code generation at build time (`icon`, `civis`, `minicss`, `Rcpp::compileAttributes()`) and of code generation at run time (`chromote`, `stevedore`). 
+Many of these examples used some form of string manipulation, in base R or with `glue`, to either generate an R script and its `roxygen2` docs **or** code using `eval()` and `parse()` (`minicss`). 
+One of them doesn't use any text representation, and `as.function` and `call`/`as.call` instead (`stevedore`).
+`icon` also doesn't write R files.
+The ability to use string manipulation
```
Comment (gaborcsardi) [inline content/post/2020-02-04-code-generation/index.md]: Unfinished sentence?

File: content/post/2020-02-04-code-generation/index.md [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+In the more general context of [automatic programming](https://en.wikipedia.org/wiki/Automatic_programming), one also finds "generative programming", and "low-code applications" (like [tidyblocks](https://tidyblocks.tech/)?).
```
Comment (gaborcsardi) [inline content/post/2020-02-04-code-generation/index.md]: one also finds them what? Unfinished sentence?

File: content/post/2020-02-04-code-generation/index.md [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+
+#### Generating icon aliases in `icon`
+
+In `icon`, an R package by Mitchell O'Hara-Wild that allows easy insertion of icons from Font Awesome, Academicons and Ionicons into R Markdown, to insert an archive icon one can type `icon::fa("archive")` or `icon::fa_archive()`, i.e. every possible icon has its own alias function which pairs well with autocompletion e.g. in RStudio when starting to type `icon::fa_`. 
+When typing `?icon::fa_archive` one gets a man page entitled "Font awesome alias", the same for all aliases. 
+How does it work?
+
+Font files related to the fonts are stored in [`inst/`](https://github.com/ropenscilabs/icon/tree/master/inst/fonts). 
+It's the case for all three fonts, but let's focus on what happens for Font Awesome. 
+Among the R code (that's executed when building the package), there's [a line reading the icon names from a font file](https://github.com/ropenscilabs/icon/blob/a5bc1cc928b15a5296b06f66faa9e08264ad4064/R/fa.R#L9).
```
[the remark this answers is not in the document]
Comment (maelle) [reply content/post/2020-02-04-code-generation/index.md]: Right I should probably say install time. But when installing from source isn't the package "built" first?

[the remark this answers is not in the document]
Comment (gaborcsardi) [reply content/post/2020-02-04-code-generation/index.md]: What is "built"? `R CMD build`? Then no.

File: content/post/2020-02-04-code-generation/index.Rmd (hunk 1 above)
Comment (maelle) [reply content/post/2020-02-04-code-generation/index.Rmd]: thanks! I copy-pasted from the PDF and hadn't caught all missing spaces :grimacing:

**Pull request merged**